sound/soc/sof/pm.c

```diff
@@ -327,12 +327,14 @@ int snd_sof_prepare(struct device *dev)
 	sdev->system_suspend_target = SOF_SUSPEND_S3;
 
 	/*
-	 * if the firmware is crashed or boot failed then we try to aim for S3
-	 * to reboot the firmware
+	 * if the firmware is crashed or boot failed then we try to aim for S4
+	 * to re-initialize the DSP and force re-load the firmware
 	 */
 	if (sdev->fw_state == SOF_FW_CRASHED ||
-	    sdev->fw_state == SOF_FW_BOOT_FAILED)
+	    sdev->fw_state == SOF_FW_BOOT_FAILED) {
+		sdev->system_suspend_target = SOF_SUSPEND_S4;
```

Comment:
I have mixed feelings about this. This is overloading a system state with side effects. It works, but can we not instead add a real flag that says 'force_firmware_boot' instead?

Reply:
@plbossart, we need to keep the S3 target in any case to make sure that the DSP is forced to be booted.

let me think about a better solution.